internal/lifecycle/lifecycle.go

```diff
@@ -74,18 +96,71 @@ func (l *Lifecycle) Start(ctx context.Context) error {
 // hooks run in reverse order.
 func (l *Lifecycle) Stop(ctx context.Context) error {
 	var errs []error
+	l.records = make(HookRecords, 0, l.numStarted)
 	// Run backward from last successful OnStart.
 	for ; l.numStarted > 0; l.numStarted-- {
 		hook := l.hooks[l.numStarted-1]
 		if hook.OnStop == nil {
 			continue
 		}
+
 		l.logger.LogEvent(&fxevent.LifecycleHookStop{CallerName: hook.caller})
+
+		l.mu.Lock()
+		l.runningHook = hook
+		l.mu.Unlock()
+
+		begin := time.Now()
 		if err := hook.OnStop(ctx); err != nil {
 			// For best-effort cleanup, keep going after errors.
 			errs = append(errs, err)
 		}
+		l.mu.Lock()
+		l.records = append(l.records, HookRecord{
+			Runtime: time.Now().Sub(begin),
+			Caller:  hook.caller,
+			Func:    hook.OnStop,
+		})
+		l.mu.Unlock()
 	}
 
 	return multierr.Combine(errs...)
 }
+
+// HookRecords returns the info of hooks that successfully ran till the end,
+// including their caller and runtime. Used to report timeout errors on Start/Stop.
+func (l *Lifecycle) HookRecords() HookRecords {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+	r := make(HookRecords, len(l.records))
+	copy(r, l.records)
+	return r
+}
+
+// RunningHookCaller returns the name of the hook that was running when a Start/Stop
+// hook timed out.
+func (l *Lifecycle) RunningHookCaller() string {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+	return l.runningHook.caller
+}
+
+// HookRecord keeps track of each Hook's execution time, the caller that appended the Hook, and function that ran as the Hook.
+type HookRecord struct {
+	Runtime time.Duration               // How long the hook ran
+	Caller  string                      // caller that appended this hook
+	Func    func(context.Context) error // function that ran as sanitized name
+}
+
+// HookRecords is a Stringer wrapper of HookRecord slice.
+type HookRecords []HookRecord
+
+// Used for logging startup errors.
+func (r HookRecords) String() string {
+	var b strings.Builder
+	sort.Slice(r, func(i, j int) bool { return r[i].Runtime > r[j].Runtime })
+	for _, r := range r {
+		b.WriteString(fmt.Sprintf("%s took %d ms to run. (Caller: %s)\n", fxreflect.FuncName(r.Func), r.Runtime.Milliseconds(), r.Caller))
```

Comment:
I'm considering the error presentation here. We have two pieces of information to report for each entry: the name of the function in the hook, and name of the function that added it.

Check out how we present information in fxreflect.Stack: https://github.com/uber-go/fx/blob/master/internal/fxreflect/stack.go#L74. We use that in error messages when a Provide or Invoke fails: the full stack trace to the function that called fx.Provide or fx.Invoke: https://github.com/uber-go/fx/blob/95337c559ed988a843a305ef45dae1b0d77ddc2f/app_test.go#L231.

I'm still only toying with presentation right now, but what do you think about multi-line formatting like that? Like the following for each entry?

```
foo/bar.Baz() took 3ms from:
    baz/qux.Whatever() (path/to/baz/qux/foo/whatever.go:42)
```

---

Separately: We don't need to truncate Runtime to milliseconds. time.Duration's default `String()` should be good enough, I think. (Use `%v` to get the default formatting for a type.)

Reply:
Thanks. I wasn't aware of the +v formatter. I tried using it and it looks much nicer so I just pushed a commit to make it look like this:

```
OnStart hook added by go.uber.org/fx_test.TestAppStart.func2.3 failed: context deadline exceeded
go.uber.org/fx_test.TestAppStart.func2.2.1() took 10.819917ms from:
    go.uber.org/fx_test.TestAppStart.func2.2 (/Users/sungyoonwhang/go/src/github.com/uber-go/fx/app_test.go:516)
go.uber.org/fx_test.TestAppStart.func2.1.1() took 291ns from:
    go.uber.org/fx_test.TestAppStart.func2.1 (/Users/sungyoonwhang/go/src/github.com/uber-go/fx/app_test.go:506)
```